### DISCUSSION TITLE
Control devices through StreamPipes&Integrate my own developed adapter module

### DISCUSSION THREAD

**Discussion by powerfulkitten** [apache/streampipes#2680] [Q&A] [answered]
Hi,I'm a new user of StreamPipes, and have some questions as follows:

1.How can I control devices through StreamPipes?
  For example, on the "PLC 4X Modbus" adapter, I can only perform readings and cannot control devices. 
  I didn't find any configuration parameters for controlling devices.

2.How can I integrate my own developed adapter module?
  For example, if I find an open-source code that integrates EtherCat using Java, how can I integrate it into StreamPipes?

Thank you very much for your help.

**Comment by bossenti** [comment, marked as answer]
Hi @powerfulkitten,

welcome to StreamPipes 👋🏼 

1. StreamPipes does not support any control operations to devices. It is restricted on reading data from PLC devices and providing processing, enrichment and analysis features for the data. In addition, you can use our sinks, such as PostgreSQL, OPC-UA, MQTT, Kafka etc., to send data to external systems.

2. Have you already discovered our [tutorial for writing custom](https://streampipes.apache.org/docs/extend-tutorial-adapters/) adapters? If so, do you miss anything?

Happy to help you more 🙂

**Reply by powerfulkitten** [reply]
Thank you, this is very helpful.

**Comment by chrisdutz** [comment]
Well ... technically PLC4X would support writing of data to PLCs ... is there any reason, this is not supported by StreamPipes? If it's supported for OPC-UA, there should not be any reason to not support writing back data to any sort of PLC.

Unfortunately we don't have EtherCat in PLC4X yet ... but it's on my ToDo list and I plan on starting to work on it as soon as I've finished the PROFINET driver.

**Reply by tenthe** [reply]
Hi @chrisdutz ,
there is no technical reason why this is not yet supported in StreamPipes. 
The only reason is that nobody has implemented it yet ;) 

I think it would be cool to have a data sink that is able to write to PLCs via PLC4X. I will create an issue for it. We should clarify a few more technical details, vut we can discuss them directly in the issue.

**Reply by powerfulkitten** [reply]
Hi guys,thanks for the response. Now I understand that PLC4X supports writing data to PLCs, but it is not yet integrated into StreamPipes (like the OPC-UA module in Data-Sinks: Writes values in an OPC-UA server). Maybe in future versions, StreamPipes will be able to support this functionality? I hope my understanding is correct.

**Reply by chrisdutz** [reply]
That's also how I understood it ... and for the PLC4X part ... yup ... that's right.

**Reply by tenthe** [reply]
Yes, you understood it correclty.
I created an issue for the PLC4X sink #2682.

If anyone is interested in working on it, please let us know. Also if you have any feedback for the functionality for such a sink, feel free to add it to the new isse.
